README


This archive contains the distribution Padre-Plugin-Mojolicious,
version 0.05:

  Simple Mojolicious helper interface for Padre

This software is copyright (c) 2010 by Breno G. de Oliveira.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
Revision history for Padre-Plugin-Mojolicious

0.05   2010.12.06
	- Restored missing POT/PO translation sources (Damyan Ivanov, AZAWAWI)

0.04   2010.11.02
	- Switched to Dist::Zilla (AZAWAWI)
	- Removed unused dependencies (AZAWAWI)
	- Code cleanup (AZAWAWI)
	- Compatibility with the latest Padre (AZAWAWI)
	- Compatibility with the latest Mojolicious (AZAWAWI)

0.03
	- Removed taint mode for tests, leading to build failure (JQUELIN)
	- General modernisation to allow it to support Padre 0.64 (ADAMK)

0.02    
	- Added translation support (SZABGAB)
	- Added translations:
	- French (jquelin)
	- Polish (THEREK)
	- Chinese (Traditional) translation (BLUET)
	- Arabic translation (AZAWAWI)
	- Spanish translation (BRUNOV)
	- Brazilian Portuguese (GARU)

0.01    2009-04-27
	- First version, released on an unsuspecting world.



lib/Padre/Plugin/Mojolicious.pm
package Padre::Plugin::Mojolicious;
BEGIN {
  $Padre::Plugin::Mojolicious::VERSION = '0.05';
}

# ABSTRACT: Simple Mojolicious helper interface for Padre

use 5.008;
use strict;
use warnings;

use Padre::Plugin ();

our @ISA = 'Padre::Plugin';

######################################################################
# Padre Integration

sub padre_interfaces {
	'Padre::Plugin' => 0.64,;
}


######################################################################
# Padre::Plugin Methods

sub plugin_name {
	Wx::gettext('Mojolicious');
}

sub plugin_disable {
	require Class::Unload;
	Class::Unload->unload('Padre::Plugin::Mojolicious::NewApp');
	Class::Unload->unload('Padre::Plugin::Mojolicious::Util');
	Class::Unload->unload('Mojolicious');
}

# The command structure to show in the Plugins menu
sub menu_plugins_simple {
	my $self = shift;
	return $self->plugin_name => [
		Wx::gettext('New Mojolicious Application') => sub {
			require Padre::Plugin::Mojolicious::NewApp;
			Padre::Plugin::Mojolicious::NewApp::on_newapp($self);
			return;
		},

		'---' => undef,

		Wx::gettext('Start Web Server') => sub {
			$self->on_start_server;
		},

		Wx::gettext('Stop Web Server') => sub {
			$self->on_stop_server;
		},

		'---' => undef,

		Wx::gettext('Mojolicious Online References') => [
			Wx::gettext('Mojolicious Manual') => sub {
				Padre::Wx::launch_browser('http://search.cpan.org/perldoc?Mojo::Manual::Mojolicious');
			},
			Wx::gettext('Mojolicious Website') => sub {
				Padre::Wx::launch_browser('http://www.mojolicious.org/');
			},
			Wx::gettext('Mojolicious Community Live Support') => sub {
				Padre::Wx::launch_irc( 'irc.perl.org' => 'mojo' );
			},

		],

		'---' => undef,

		Wx::gettext('About') => sub {
			$self->on_show_about;
		},
	];
}

sub on_start_server {
	my $self = shift;
	my $main = $self->main;

	require Padre::Plugin::Mojolicious::Util;
	my $project_dir = Padre::Plugin::Mojolicious::Util::get_document_base_dir();

	my $server_filename = Padre::Plugin::Mojolicious::Util::get_mojolicious_project_name($project_dir);

	my $server_full_path = File::Spec->catfile( $project_dir, 'script', $server_filename );
	unless ( -e $server_full_path ) {
		Wx::MessageBox(
			sprintf(
				Wx::gettext(
					"Mojolicious application script not found at\n%s\n\nPlease make sure the active document is from your Mojolicious project."
				),
				$server_full_path
			),
			Wx::gettext('Server not found'),
			Wx::wxOK, $main
		);
		return;
	}

	# Go to the selected file's directory
	# (Mojolicious instructs us to always run their scripts
	#  from the basedir)
	my $pwd = Cwd::cwd();
	chdir $project_dir;

	require Padre::Perl;
	my $command = Padre::Perl->cperl . ' ' . File::Spec->catfile( 'script', $server_filename ) . ' daemon';

	$main->run_command($command);

	# restore current dir
	chdir $pwd;

	# TODO: actually check whether this is true.
	my $ret = Wx::MessageBox(
		Wx::gettext('Web server appears to be running. Launch web browser now?'),
		Wx::gettext('Start Web Browser?'),
		Wx::wxYES_NO | Wx::wxCENTRE,
		$main,
	);
	if ( $ret == Wx::wxYES ) {
		Padre::Wx::launch_browser('http://localhost:3000');
	}

	# TODO: handle menu greying

	return;
}

sub on_stop_server {
	my $self = shift;
	my $main = $self->main;

	# TODO: Make this actually call
	# Run -> Stop
	if ( $main->{command} ) {
		my $processid = $main->{command}->GetProcessId();
		kill( 9, $processid );

		# $main->{command}->TerminateProcess;
	}
	delete $main->{command};
	$main->menu->run->enable;
	$main->output->AppendText( Wx::gettext("\nWeb server stopped successfully.\n") );
	return;
}

sub on_show_about {
	require Mojolicious;
	require Class::Unload;
	my $about = Wx::AboutDialogInfo->new;
	$about->SetName("Padre::Plugin::Mojolicious");
	$about->SetDescription( "Initial Mojolicious support for Padre\n\n"
			. "This system is running Mojolicious version "
			. $Mojolicious::VERSION
			. "\n" );
	$about->SetVersion($Padre::Plugin::Mojolicious::VERSION);
	Class::Unload->unload('Mojolicious');
	Wx::AboutBox($about);
	return;
}


1;



=pod

=head1 NAME

Padre::Plugin::Mojolicious - Simple Mojolicious helper interface for Padre

=head1 VERSION

version 0.05

=head1 SYNOPSIS

	cpan install Padre::Plugin::Mojolicious;

Then use it via L<Padre>, The Perl IDE.

=head1 DESCRIPTION

Once you enable this Plugin under Padre, you'll get a brand new menu with the following options:

=head2 'New Mojolicious Application'

This options lets you create a new Mojolicious application.

=head2 'Start Web Server'

This option will automatically spawn your application's development web server. Once it's started, it will ask to open your default web browser to view your application running.

Note that this works like Padre's "run" menu option, so any other execution it will be disabled while your server is running.

=head2 'Stop Web Server'

This option will stop the development web server for you.

=head2 'Mojolicious Online References'

This menu option contains a series of external reference links on Mojolicious. Clicking on each of them will point your default web browser to their websites.

=head2 'About'

Shows a nice about box with this module's name and version.

=head1 BUGS

Please report any bugs or feature requests to C<bug-padre-plugin-mojolicious at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Padre-Plugin-Mojolicious>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

	perldoc Padre::Plugin::Mojolicious

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Padre-Plugin-Mojolicious>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Padre-Plugin-Mojolicious>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Padre-Plugin-Mojolicious>

=item * Search CPAN

L<http://search.cpan.org/dist/Padre-Plugin-Mojolicious/>

=back

=head1 SEE ALSO

L<Mojolicious>, L<Padre>

=head1 AUTHORS

=over 4

=item *

Breno G. de Oliveira <garu@cpan.org>

=item *

Ahmad M. Zawawi <ahmad.zawawi@gmail.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Breno G. de Oliveira.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


__END__



lib/Padre/Plugin/Mojolicious/Util.pm
package Padre::Plugin::Mojolicious::Util;
BEGIN {
  $Padre::Plugin::Mojolicious::Util::VERSION = '0.05';
}

# ABSTRACT: Some code used all around the Plugin

use 5.008;
use strict;
use warnings;
use Cwd            ();
use File::Spec     ();
use File::Basename ();

# get the Mojolicious project name, so we can
# figure out the development server's name
# TODO: make this code suck less
sub get_mojolicious_project_name {
	my $project_dir = shift;
	return unless $project_dir;

	require File::Spec;
	my @dirs         = File::Spec->splitdir($project_dir);
	my $project_name = lc( $dirs[-1] );
	$project_name =~ tr{-}{_};

	return $project_name;
}

sub find_file_from_output {
	my $filename    = shift;
	my $output_text = shift;

	$filename .= '.pm';

	if ( $output_text =~ m{\[write\] (.+$filename)} ) {
		return $1;
	} else {
		return; # sorry, not found
	}
}

sub get_document_base_dir {
	my $main     = Padre->ide->wx->main;
	my $doc      = $main->current->document;
	my $filename = $doc->filename;

	return unless $filename;

	# check for potential relative path on filename
	if ( $filename =~ m{\.\.} ) {
		require Cwd;
		$filename = Cwd::realpath($filename);
	}
	my $olddir = File::Basename::dirname($filename);
	my $dir    = $olddir;
	while (1) {

		# those are the test cases for a Mojolicious directory
		if (   -d File::Spec->catfile( $dir, 'script' )
			&& -d File::Spec->catfile( $dir, 'lib' )
			&& -d File::Spec->catfile( $dir, 'log' )
			&& -d File::Spec->catfile( $dir, 'public' )
			&& -d File::Spec->catfile( $dir, 't' )
			&& -d File::Spec->catfile( $dir, 'templates' ) )
		{
			return $dir;
		}
		$olddir = $dir;
		$dir    = File::Basename::dirname($dir);

		last if $olddir eq $dir;
	}
	return;
}

1;

__END__
=pod

=head1 NAME

Padre::Plugin::Mojolicious::Util - Some code used all around the Plugin

=head1 VERSION

version 0.05

=head1 AUTHORS

=over 4

=item *

Breno G. de Oliveira <garu@cpan.org>

=item *

Ahmad M. Zawawi <ahmad.zawawi@gmail.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Breno G. de Oliveira.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Padre/Plugin/Mojolicious/NewApp.pm
package Padre::Plugin::Mojolicious::NewApp;
BEGIN {
  $Padre::Plugin::Mojolicious::NewApp::VERSION = '0.05';
}

# ABSTRACT: A New Mojolicious Application for Padre

use 5.008;
use strict;
use warnings;
use Cwd               ();
use File::Spec        ();
use Padre::Wx         ();
use Padre::Wx::Dialog ();
use Padre::DB         ();

sub on_newapp {
	my $plugin = shift;
	my $main   = $plugin->main;
	my $dialog = dialog($main);
	$dialog->Show(1);
	return;
}

sub get_layout {

	my @layout = (
		[   [ 'Wx::StaticText', undef,        Wx::gettext('Application Name:') ],
			[ 'Wx::TextCtrl',   '_app_name_', '' ],
		],
		[   [ 'Wx::StaticText', undef, Wx::gettext('Parent Directory:') ],
			[ 'Wx::DirPickerCtrl', '_directory_', '', Wx::gettext('Pick parent directory') ],
		],
		[   [ 'Wx::Button', '_ok_',     Wx::wxID_OK ],
			[ 'Wx::Button', '_cancel_', Wx::wxID_CANCEL ],
		],
	);
	return \@layout;
}

sub dialog {
	my $parent = shift;
	my $config = $parent->current->config;
	my $layout = get_layout();
	my $dialog = Padre::Wx::Dialog->new(
		parent => $parent,
		title  => Wx::gettext('New Mojolicious Application'),
		layout => $layout,
		width  => [ 100, 200 ],
		bottom => 20,
	);

	$dialog->{_widgets_}->{_directory_}->SetPath( $config->module_start_directory );

	$dialog->{_widgets_}->{_ok_}->SetDefault;
	Wx::Event::EVT_BUTTON( $dialog, $dialog->{_widgets_}->{_ok_}, \&ok_clicked );

	Wx::Event::EVT_BUTTON( $dialog, $dialog->{_widgets_}->{_cancel_}, \&cancel_clicked );

	$dialog->{_widgets_}->{_app_name_}->SetFocus;

	return $dialog;
}


sub cancel_clicked {
	my $dialog = shift;
	$dialog->Destroy;
	return;
}

sub ok_clicked {
	my $dialog = shift;
	my $event  = shift;
	my $data   = $dialog->get_data;
	$dialog->Destroy;

	my $main = $dialog->GetParent->current->main;

	# TODO improve input validation !
	if ( $data->{'_app_name_'} =~ m{^\s*$|[^\w\:]}o ) {
		Wx::MessageBox( Wx::gettext('Invalid Application name'), Wx::gettext('missing field'), Wx::wxOK, $main );
		return;
	} elsif ( not $data->{'_directory_'} ) {
		Wx::MessageBox(
			Wx::gettext('You need to select a base directory'), Wx::gettext('missing field'), Wx::wxOK,
			$main
		);
		return;
	}

	# We should probably call Mojolicious::Helper directly
	# (new() and mk_app()) here, as long as we can redirect
	# print statements to $main->output->AppendText().
	#
	# Perhaps if run_command() were to block before continuing,
	# we could use something like:
	#$main->run_command('mojolicious generate app' . $data->{'_app_name_'});

	# Prepare the output window for the output
	$main->show_output(1);
	$main->output->Remove( 0, $main->output->GetLastPosition );

	my @command = (
		'mojo',
		'generate',
		'app',
		$data->{'_app_name_'},
	);

	# go to the selected directory
	my $pwd = Cwd::cwd();
	chdir $data->{'_directory_'};

	# run command, then immediately restore directory
	my $output_text = qx(@command);
	chdir $pwd;

	$main->output->AppendText($output_text);

	my $ret = Wx::MessageBox(
		sprintf( Wx::gettext("%s apparently created. Do you want to open it now?"), $data->{_app_name_} ),
		Wx::gettext('Done'),
		Wx::wxYES_NO | Wx::wxCENTRE,
		$main,
	);
	if ( $ret == Wx::wxYES ) {
		require Padre::Plugin::Mojolicious::Util;
		my $file = Padre::Plugin::Mojolicious::Util::find_file_from_output(
			$data->{'_app_name_'},
			$output_text
		);
		$file = Cwd::realpath($file); # avoid relative paths

		Padre::DB::History->create(
			type => 'files',
			name => $file,
		);
		$main->setup_editor($file);
		$main->refresh;
	}

	return;
}

1;

__END__
=pod

=head1 NAME

Padre::Plugin::Mojolicious::NewApp - A New Mojolicious Application for Padre

=head1 VERSION

version 0.05

=head1 AUTHORS

=over 4

=item *

Breno G. de Oliveira <garu@cpan.org>

=item *

Ahmad M. Zawawi <ahmad.zawawi@gmail.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Breno G. de Oliveira.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  'build_requires' => {
    'Locale::Msgfmt' => '0.15',
    'Module::Build' => '0.3601',
    'Test::More' => '0'
  },
  'configure_requires' => {
    'Module::Build' => '0.3601'
  },
  'dist_abstract' => 'Simple Mojolicious helper interface for Padre',
  'dist_author' => [
    'Breno G. de Oliveira <garu@cpan.org>',
    'Ahmad M. Zawawi <ahmad.zawawi@gmail.com>'
  ],
  'dist_name' => 'Padre-Plugin-Mojolicious',
  'dist_version' => '0.05',
  'license' => 'perl',
  'module_name' => 'Padre::Plugin::Mojolicious',
  'recommends' => {},
  'recursive_test_files' => 1,
  'requires' => {
    'File::Basename' => '0',
    'Mojolicious' => '0.999934',
    'Padre' => '0.64'
  },
  'script_files' => [],
  'share_dir' => {
    'dist' => 'share'
  }
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
